### PR TITLE
Fixes Issues 2865, 2866, 2870, 2871 in CriticalPath AsyncCSS and DOM Module

### LINKED ISSUE #2865: 3.6.2: Fatal Error Thrown When HTMLDocument::query invoked but XPath not set up
See ticket https://secure.helpscout.net/conversation/1220180136/179215/

A fatal error is thrown when XPath is not set up and `HTMLDocument::query` is invoked. 

```
PHP Fatal error: Uncaught Error: Call to a member function query() on bool in /var/www/x/public_html/x/wp-content/plugins/wp-rocket/inc/Engine/CriticalPath/AsyncCSS.php:134

Stack trace:
#0 /var/www/x/public_html/x/wp-content/plugins/wp-rocket/inc/Engine/CriticalPath/CriticalCSSSubscriber.php(591): WP_Rocket\\Engine\\CriticalPath\\AsyncCSS->modify_html()
#1 /var/www/x/public_html/x/wp-includes/class-wp-hook.php(287): WP_Rocket\\Engine\\CriticalPath\\CriticalCSSSubscriber->async_css()
#2 /var/www/x/public_html/x/wp-includes/plugin.php(206): WP_Hook->apply_filters()
#3 /var/www/x/public_html/x/wp-content/plugins/wp-rocket/x/classes/Buffer/class-optimization.php(104): apply_filters()
#4 [internal function]: WP_Rocket\\Buffer\\Optimization->maybe_process_buffer()
#5 /var/www/x/public_html/x/wp-includes/functions.php(4609): ob_end_flush()
#6 /var/www/hoke.org/public_html/john/wp-includes/class-wp-hook.php(287): wp_ob_end_flush_all()
#7 ...
```

This appears to happen if an empty string is passed to one of the named constructors. However, we should still guard against this to ensure no fatal error is thrown.

### LINKED ISSUE #2866: 3.6.2 Encoding issues with DOM Module for Non-Latin Characters causing broken layouts
`DOMDocument` encoding sets the `<meta>` tag but not the overall HTML encoding. When creating the DOM and saving it back to a string, the UTF-8 encoding is not respected. As such, it creates encoded entities and in some cases can break the layout.

### LINKED ISSUE #2870: 3.6.2 PHP Warning when passing an empty HTML string to DOM Module
See ticket https://secure.helpscout.net/conversation/1220180136/179215/

When an empty string is passed to the DOM Module, a PHP Warning is generated.:

```
PHP Warning: DOMDocument::loadHTML(): Empty string supplied as input in /var/www/xxx.xxx/public_html/xxx/wp-content/plugins/wp-rocket/inc/Engine/DOM/HTMLDocument.php on line 193
```

Solution: 

- Add guard clauses to bail out when an empty is received.

### LINKED ISSUE #2871: 3.6.2 LazyLoad and Optimize CSS Delivery break layouts when both are enabled
When LazyLoad and Optimize CSS Delivery options are both enabled in 3.6.2, it breaks the layout when there's malformed HTML.

We validated this on @ScottTravisHartley's website. https://secure.helpscout.net/conversation/1220716013/179344

**To Reproduce**
1. Use @ScottTravisHartley's website
2. Turn on Optimize CSS Delivery and LazyLoad. Save.
3. View the site incognito
4. Notice the images are not rendering correctly and the layout is broken.

Next, turn off LazyLoad (images and iframe/videos). Save. Critical CSS generates. Refresh the main page and the site renders correctly. 

Turn off Optimize CSS Delivery and turn on LazyLoad. Save. Refresh. Site renders correctly.

**Expected behavior**
We expect the site to render correctly when one or both options are turned on.

**Additional context**

When both are off:

```html
<div class="mainVisualCover" style="background: url('https://xxxx/assets/uploads/2017/01/Top-Page-Video-Logos-hrtech2018.png') no-repeat left center; background-size: cover;"></div>
<img src="https://xxxx/assets/uploads/2017/01/main_visual_sp-1024.png" alt="Main Visual">
<video src="https://xxxx/assets/uploads/2017/01/movie.mp4" autoplay="" preload="auto" poster="https://xxxx/assets/uploads/2017/01/main_visual.png"></video>
```

LazyLoad is on and Optimize CSS Delivery is on:
```html
<div class="mainVisualCover" style="background: url('https://xxxx/assets/uploads/2017/01/Top-Page-Video-Logos-hrtech2018.png') no-repeat left center; background-size: cover;"></div>
<img src="https://xxxx/assets/uploads/2017/01/main_visual_sp-1024.png" alt="Main Visual" class="lazyloaded" data-ll-status="loaded">
<noscript><img src="https://xxxx/assets/uploads/2017/01/main_visual_sp-1024.png" alt="Main Visual"></noscript>
<video src="https://xxxx/assets/uploads/2017/01/movie.mp4" autoplay="" preload="auto" poster="https://xxxx/assets/uploads/2017/01/main_visual.png"></video>
```

LazyLoad is off and Optimize CSS Delivery is on:
```html
<div class="mainVisualCover" style="background: url('https://xxxx/assets/uploads/2017/01/Top-Page-Video-Logos-hrtech2018.png') no-repeat left center; background-size: cover;"></div>
<img src="https://xxxx/assets/uploads/2017/01/main_visual_sp-1024.png" alt="Main Visual">
<video src="https://xxxx/assets/uploads/2017/01/movie.mp4" autoplay="" preload="auto" poster="https://xxxx/assets/uploads/2017/01/main_visual.png"></video>
```

LazyLoad is on and Optimize CSS Delivery is off:
```html
<div class="mainVisualCover" style="background: url('https://xxxx/assets/uploads/2017/01/Top-Page-Video-Logos-hrtech2018.png') no-repeat left center; background-size: cover;"></div>
<img src="data:image/svg+xml,%3Csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%200%200'%3E%3C/svg%3E" alt="Main Visual" data-lazy-src="https://xxxx/assets/uploads/2017/01/main_visual_sp-1024.png">
<noscript><img src="https://xxxx/assets/uploads/2017/01/main_visual_sp-1024.png" alt="Main Visual"></noscript>
<video src="https://xxxx/assets/uploads/2017/01/movie.mp4" autoplay="" preload="auto" poster="https://xxxx/assets/uploads/2017/01/main_visual.png"></video>
```

Notice the differences. These are clues.

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### LINKED ISSUE #2885: 3.6.2: Include media attribute when loading CSS files async
This is related to the new method for loading CSS files asynchronously with the Optimize CSS Delivery option, introduced in 3.6.2.

We need to account for the @media attribute when we load the CSS files. 

Related ticket: https://secure.helpscout.net/conversation/1220486412/179297?folderId=377611

Here the customer is using the Avada theme. Many of their files are using specific media attributes, e.g.: 
`media='only screen and (min-width: 784px) and (max-width: 856px)'`

When these are preloaded without the correct media attribute, it causes additional HTTP requests. 
Problem can be seen in these test results: https://secure.helpscout.net/conversation/1220486412/179297?folderId=377611#thread-3499389455

Per @hellofromtonya :

> For each stylesheet, we now do the following:
> Add a `<link rel="stylesheet" href="" as="style">`   - This doesn’t include the @media attribute.
> Modify the original <link> ’s onload and media to make it async.
> The first one needs the media attribute too.

### PR DESCRIPTION
**Pull request by hellofromtonya**
Closes #2865 
Closes #2866 
Closes #2870
Closes #2871 

Closes #2885

Adds the updated DOM Module and its test cases to fix the following issues:

- [x] Bugfix: character encoding 
     - [x] improves encoding to handle non-Latin characters (different languages added to the test cases to validate)
     - [x] Auto detect character encoding - see [DOM Module issue 21](https://github.com/wp-media/module-dom/issues/21)
     - [x] Preserve HTML entities after processing - see [DOM Module Issue 22](https://github.com/wp-media/module-dom/issues/22)
- [x] Bugfix: Adds guard clauses to prevent PHP notice when an empty string is passed into `HTMLDocument::from_html`, `HTMLDocument::from_fragment`, and `HTMLDocument:loadHtml`.
- [x] Bugfix: Adds guard clause to `HTMLDocument::query` to ensure XPath is set up
- [x] Adds test cases to validate that the `type` attribute is not being stripped from `<script>` element
- [x] Bugfix: Preserves HTML templates inside `<script>` elements
- [x] Preserves malformed and invalid content inside of the `<head></head>` 
- [x] Preserves`<noscript>` in `<head>` when libxml is < 2.8.0
- [x] Attempts to fix/normalize the basic HTML structure of
```html
<!DOCTYPE html>
<html>
<head>
</head>
<body>
</body>
</html>
```
- [x] Transforms self-closing tags to HTML5 format, e.g. `<br/>`, `<br />` , `<br         />` etc are transformed into `<br>`
- [x] Regression: include media attribute when loading CSS files async Issue #2885

Credit props to @engahmeds3ed for identifying the encoding issue with non-Latin characters 🥇

### CODE REVIEW

**Review by remyperona — APPROVED** [review]

**Comment by hellofromtonya** [comment]
This PR is ready for @wp-media/qa 

Things to check in this one:

1. Add non-Latin characters or language to a web page and check that the characters remain intact. For example, [check out this test data in the fixture](https://secure.helpscout.net/conversation/1219841185/179108).
2. Testing an empty string's impact can really only be done by [modifying the code here](https://github.com/wp-media/wp-rocket/blob/updates/dom-module-for-fixes/inc/Engine/CriticalPath/CriticalCSSSubscriber.php#L591) to pass an `''` empty string instead of `$html`.

**Review by Davidacu — APPROVED** [review]
Testing Environment:
http://mega.wp-rocket.me/qa/

1. I can confirm that non-Latin characters are being handled properly now.
Before in 3.6.2 -> https://i.imgur.com/yTk9ZJW.jpg
Now -> https://i.imgur.com/2NJmtH4.jpg

2. I manually modified the [code here](https://github.com/wp-media/wp-rocket/blob/updates/dom-module-for-fixes/inc/Engine/CriticalPath/CriticalCSSSubscriber.php#L591) to pass an empty string and did not generate any PHP warnings on Critical CSS regeneration.

3. Confirmed the `<link>` elements inside `<noscript>` at the bottom are valid elements, chrome does show them as a string if JS is enabled, but if JS is disabled it does show as elements.

**Comment by hellofromtonya** [comment]
This PR has been updated. Let's review what's new and changed since the last QA and code review:

Websites are filled with invalid and malformed HTML/content. Before DOM Module, Rocket didn't care if the HTML was well-formed or not. With `DOMDocument`, the DOM Module needs to do more to be flexible to work with invalid and malformed HTML/content.

The new DOM Module (bundled in this PR) is now more flexible. 

- Preserves invalid and/or malformed HTML/content in the `<head>`. Issue #2871
- Adds missing basic HTML structure tags when missing: Issue #2871
    - Adds `<DOCTYPE>`, `<html>`, `<head>`, `</head>`, `<body>`, `</body>`, `</html>`
    - However, if the `</head>` and `<body>` tags are missing, we bail out. Why? We can't guess where those should be inserted. If both are missing, then AsyncCSS (Optimized CSS Delivery) is not applied to the web page.
- Preserves closing tags contained with `<script>`, i.e. for JS-powered templates to insert HTML.
- Preserves all `<noscript>` elements in the `<head>`, even if the hosting Libxml version is < 2.8.0. Issue #2871
- Preserves the charset encoding for non-Latin characters, i.e. Issue #2866
- And retains the fixes for Issues #2865 and #2870

**Comment by hellofromtonya** [comment]
@wp-media/qa This PR is ready for re-QA.

File: inc/Engine/DOM/Transformer/Normalizer.php [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+	 * Information about the <body> tag.
+	 *
+	 * @var array|bool
+	 */
+	private $body_opening_tag;
+
+	/**
+	 * Normalize the document structure.
+	 *
+	 * This makes sure the document adheres to the general structure that AMP requires:
```
Comment (Tabrisrp) [inline inc/Engine/DOM/Transformer/Normalizer.php:52]:
```suggestion
	 * This makes sure the document adheres to the general structure that WP Rocket requires:
```

Comment (hellofromtonya) [reply inc/Engine/DOM/Transformer/Normalizer.php]: Resolved https://github.com/wp-media/wp-rocket/pull/2868/commits/b769cd142e8545dbbcad4d476072f5993763db55

**Review by remyperona — APPROVED** [review]

**Review by iCaspar — APPROVED** [review]

**Comment by hellofromtonya** [comment]
Update on this PR:

After testing on multiple customer sites, we found 1:30 had missing `</div>` tags. These types of unclosed tags break sites as `DOMDocument` is unable to identify the exact `<div>` that was not closed. Why? `<div>` can be nested.

Knowing that this is a breaking issue, we decided to add a new feature to the DOM Module: apply only changes made to the DOM to the actual HTML.

How is this different?

Instead of converting the entire DOM back into a HTML string, we will only apply mapped changes made to the string. In this way, invalid tags, malformed HTML, and/or unclosed tags are not edited or changed, i.e. the original HTML for these remains.

I'm closing this PR and will open a new PR to capture the new enhancements.

**Pull request closed without merging**